lib/_stream_transform.js

```diff
@@ -217,3 +230,110 @@ function done(stream, er, data) {
   return stream.push(null);
 }
+
+function SourceIterator(asyncGeneratorFn, opts) {
+  const source = this;
+  this[kSourceIteratorPull] = null;
+  this[kSourceIteratorChunk] = null;
+  this[kSourceIteratorResolve] = null;
+  this[kSourceIteratorGenerator] = asyncGeneratorFn;
+  this[kSourceIteratorStream] = new Transform({
+    objectMode: true,
+    ...opts,
+    transform(chunk, encoding, cb) {
+      source.encoding = encoding ||
+        source[kSourceIteratorStream]._transformState.writeencoding;
+      if (source[kSourceIteratorResolve] === null) {
+        source[kSourceIteratorChunk] = chunk;
+        source[kSourceIteratorPull] = cb;
+        return;
+      }
+      source[kSourceIteratorResolve]({ value: chunk, done: false });
+      source[kSourceIteratorResolve] = null;
+      cb(null);
+    }
+  });
+  this.encoding = this[kSourceIteratorStream]._transformState.writeencoding;
+  this[kSourceIteratorGrabResolve] = (resolve) => {
+    this[kSourceIteratorResolve] = resolve;
+  };
+  this[kSourceIteratorPump]();
+}
+
+SourceIterator.prototype[Symbol.asyncIterator] = function() {
+  return this;
+};
+
+Object.setPrototypeOf(SourceIterator.prototype,
+                      AsyncIteratorPrototype);
+
+SourceIterator.prototype.next = function next() {
+  if (this[kSourceIteratorPull] === null || this[kSourceIteratorChunk] === null)
+    return new Promise(this[kSourceIteratorGrabResolve]);
+
+  this[kSourceIteratorPull](null);
+  const result = Promise.resolve({
+    value: this[kSourceIteratorChunk],
+    done: false });
+  this[kSourceIteratorChunk] = null;
+  this[kSourceIteratorPull] = null;
+  return result;
+};
+
+SourceIterator.prototype[kSourceIteratorPump] = async function pump() {
+  const stream = this[kSourceIteratorStream];
+  try {
+    const asyncGeneratorFn = this[kSourceIteratorGenerator];
+    const transformIter = asyncGeneratorFn(this);
+    stream.removeListener('prefinish', prefinish);
+    stream.on('prefinish', () => {
+      if (this[kSourceIteratorResolve] !== null) {
+        this[kSourceIteratorResolve]({ value: undefined, done: true });
+      }
+    });
+    let next = await transformIter.next();
+    do {
+      if (next.done) {
+        if (next.value !== undefined) stream.push(next.value);
+
+        // In the event of an early return we explicitly
+        // discard any buffered state
+        if (stream._writableState.length > 0) {
+          const { length } = stream._writableState;
+          const { transforming } = stream._transformState;
+          stream._writableState.length = 0;
+          stream._transformState.transforming = false;
+          prefinish.call(stream);
+          stream._writableState.length = length;
+          stream._transformState.transforming = transforming;
+        } else {
+          prefinish.call(stream);
+        }
+        break;
+      }
+      stream.push(next.value);
+    } while (next = await transformIter.next());
```

Comment:
next() will never return a falsy value, you might want to consider using a different loop condition to make the logic clearer.

Reply:
I could change it to `while(true)` but I like the way the `await` is in the condition, it shows each loop is async..
would an explanatory comment satisfy?

Reply:
Ideally either
```js
while (true) {
  const { value, done } = await transformIter.next();
  if (done) {
    // xyz
    break;
  }
  // abc
}
```
or
```js
for await (const value of transformIter) {
  // abc
}
// xyz
```

Reply:
the second case only handles yielded values not return values,  (otherwise I would have used it)

I can change to the first form

Reply:
I'm not convinced we should treat the return value as a yielded one. It's a bit unintuitive IMO.

Reply:
I think its unintuitive not to 🤣 

Consider this scenario:

```js
 async function * mapper(source) {
    for await (const chunk of source) {
      yield doSomething(chunk)
    }
    return 'last chunk' // equivalent to stream.end('last chunk')
 }
```

If we *don't* treat return as a yielded value, any returned value is ignored. In no other cases is a return value in a function ignored, so ignoring it would violate the principal of least surprise. 

Here's another scenario: returning from within a loop. 

Let's start with a synchronous analogue:

```js
function find () {
  for (const value of array) {
     if (matchesCriteria(value)) return value
  }
}
```

This is valid, and I'd argue fairly common.

Now an async function version:

```js
async function find () {
  for (const value of array) {
     if (await matchesCriteria(value)) return value
  }
}
```

Again totally valid and clear, now let's map it to an async function generator passed to `Transform.by`:

```js
  async function * endStreamWhenValueFound(source) {
    for await (const chunk of source) {
      yield doSomething(chunk) // maybe propagate stats or something
      if (await matchesCriteria(value)) return chunk // value found, end stream with value
    }
  }
```

Reply:
@targos something i just learned last week is that `yield*` includes the last `value` with the `done: true`, so it actually kinda does fit to include the last value.

Reply:
I see @targos +1 on @devsnek last comment - resolving